test_concurrent_specimen_submissions fails intermittently

Here is the stack track:
```
======================================================================
FAIL: test_concurrent_specimen_submissions (indexer.test_hca_indexer.TestHCAIndexer)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/DataBiosphere/azul/test/indexer/test_hca_indexer.py", line 231, in test_concurrent_specimen_submissions
    self.assertEqual(num_hits, 2)
AssertionError: 1 != 2
----------------------------------------------------------------------
```

Test passed locally and then again once the Travis build reran.

Not sure if this is a real issue or not but I thought I'd report.

┆Issue is synchronized with this [Jira Story](https://ucsc-cgl.atlassian.net/browse/AZUL-431)
┆Project Name: azul
┆Issue Number: AZUL-431

--- a/src/azul/threads.py
+++ b/src/azul/threads.py
@@ -1,0 +1,76 @@
+import threading
+
+from azul import require
+
+
+class Latch:
+    """
+    >>> l = Latch(1)
+    >>> l.decrement(1)  # opens the latch
+    >>> l.decrement(1)  # latch already open, but decrements value
+    >>> l.value
+    -1
+
+    >>> l = Latch(0)  # latch is initially open
+    >>> l.decrement(1)
+    >>> l.value
+    -1
+
+    >>> l = Latch(0)
+    >>> l.decrement(0)
+    >>> l.value
+    0
+
+
+    >>> l = Latch(value=0)
+    >>> l.decrement(0, timeout=0.01)
+    >>> l.value
+    0
+
+    >>> l = Latch(value=2)
+    >>> l.decrement(1, timeout=0.01)  # not enough to open latch, so time out
+    Traceback (most recent call last):
+    ...
+    TimeoutError
+    >>> l.decrement(1)  # opens latch
+    >>> l.value
+    0
+
+    >>> from concurrent.futures import ThreadPoolExecutor
+    >>> n = 2
+    >>> with ThreadPoolExecutor(max_workers=n) as tpe:
+    ...     l = Latch(n)
+    ...     fs = [tpe.submit(l.decrement, 1) for i in range(n)]
+    >>> [f.result() for f in fs]
+    [None, None]
+
+    >>> with ThreadPoolExecutor(max_workers=n) as tpe:
+    ...     l = Latch(n+1)
+    ...     fs = [tpe.submit(l.decrement, 1, timeout=1) for i in range(n)]
+    >>> [f.result() for f in fs]
+    Traceback (most recent call last):
+    ...
+    TimeoutError
+    """
+
+    def __init__(self, value):
+        require(isinstance(value, int))
+        self.value = value
+        self.condition = threading.Condition()
+
+    def decrement(self, value, timeout=None):
+        require(isinstance(value, int))
+        self.condition.acquire()
+        try:
+            self.value -= value
+            if self.value > 0:
+                while True:
+                    if self.condition.wait(timeout=timeout):
+                        if self.value <= 0:
+                            break
+                    else:
+                        raise TimeoutError()
+            else:
+                self.condition.notifyAll()
+        finally:
+            self.condition.release()

--- a/test/indexer/test_hca_indexer.py
+++ b/test/indexer/test_hca_indexer.py
@@ -9,6 +9,7 @@ from elasticsearch import Elasticsearch
 from more_itertools import one
 
 from azul import config
+from azul.threads import Latch
 from azul.transformer import Aggregate, Contribution
 from indexer import IndexerTestCase
 
@@ -209,11 +210,12 @@ class TestHCAIndexer(IndexerTestCase):
         bundles = [("9dec1bd6-ced8-448a-8e45-1fc7846d8995", "2018-03-29T154319.834528Z"),
                    ("56a338fe-7554-4b5d-96a2-7df127a7640b", "2018-03-29T153507.198365Z")]
         original_mget = Elasticsearch.mget
+        latch = Latch(len(bundles))
 
         def mocked_mget(self, body):
             mget_return = original_mget(self, body=body)
-            # both threads sleep after reading to force conflict while writing
-            time.sleep(0.5)
+            # all threads wait at the latch after reading to force conflict while writing
+            latch.decrement(1)
             return mget_return
 
         with patch.object(Elasticsearch, 'mget', new=mocked_mget):

--- a/test/test_doctests.py
+++ b/test/test_doctests.py
@@ -4,6 +4,7 @@ import unittest
 import azul
 import azul.json_freeze
 from azul.modules import load_module
+import azul.threads
 import azul.time
 import azul.transformer
 import azul.vendored.frozendict
@@ -11,6 +12,7 @@ import azul.vendored.frozendict
 
 def load_tests(loader, tests, ignore):
     tests.addTests(doctest.DocTestSuite(azul))
+    tests.addTests(doctest.DocTestSuite(azul.threads))
     tests.addTests(doctest.DocTestSuite(azul.time))
     tests.addTests(doctest.DocTestSuite(azul.transformer))
     tests.addTests(doctest.DocTestSuite(azul.json_freeze))

